Tweaks after testing modules.

ajaypanyala opened the pull request.

ajaypanyala committed:
Tweaks after testing modules

ajaypanyala committed:
fix

wavefunction91 commented on FindBLIS.cmake:
```diff
@@ -9,6 +9,8 @@ else()
   set( BLIS_LIBRARY_NAME "blis" )
 endif()
 
+set( BLIS_PREFIX ${BLISROOT} $ENV{BLISROOT} )
```
This should be guarded to avoid overriding user input
```
if( NOT DEFINED BLIS_PREFIX )
  set( BLIS_PREFIX ${BLISROOT} $ENV{BLISROOT} )
endif()
```

ajaypanyala replied:
Fixed!

wavefunction91 commented on FindBLIS.cmake:
```diff
@@ -25,11 +27,10 @@ find_path( BLIS_INCLUDE_DIR
+if( BLIS_LIBRARIES )
+ find_package( Threads QUIET )
+ set( BLIS_LIBRARIES ${BLIS_LIBRARIES} Threads::Threads "m")
+endif()
```
These shouldn't be needed, I have checks for lm in `FindBLAS`, I can add threads too

ajaypanyala committed:
check BLIS_PREFIX

ajaypanyala committed:
fix prefix

ajaypanyala committed:
missed Ref.BLAS prefix

ajaypanyala committed:
throw error when ilp64 interface is requested for ScaLAPACK discovery

ajaypanyala committed:
Add Fortran compile flags for Flang

ajaypanyala committed:
mkl_sycl discovery cleanup

ajaypanyala committed:
fix

ajaypanyala committed:
Fix for Threads discovery on MACOS

wavefunction91 committed:
Hardcode macro installation paths

ajaypanyala committed:
revert

ajaypanyala committed:
remove ROOT vars

wavefunction91 commented on FindBLAS.cmake:
```diff
@@ -49,8 +49,8 @@ if( NOT BLAS_LIBRARIES )
-      set( BLAS_INCLUDE_DIR         "${${blas_type}_INCLUDE_DIR}"         )
-
+      set( BLAS_INCLUDE_DIRS        "${${blas_type}_INCLUDE_DIR}"         )
+      set( BLAS_COMPILE_OPTIONS     "${${blas_type}_C_COMPILE_FLAGS}"     )
```
This brings up an interesting point, we may need to separate out `C_COMPILE_OPTIONS`, etc for every language (with the default being `C`=`CXX`). Eventually we may want to create separate `TARGET`'s, e.g. `BLAS::BLAS_C`, `BLAS::BLAS_Fortran`, I'll make an issue for it.

The only think I'd change here is to avoid specifying language until we can propagate it consistently, i.e. `C_COMPILE_FLAGS` -> `COMPILE_OPTIONS`

wavefunction91 commented on FindIntelMKL.cmake:
```diff
@@ -253,6 +253,16 @@ else()
 endif()
 
+# SYCL
+if(ENABLE_DPCPP)
```
SYCL bindings =/= DPCPP. The way I'd go about this is adding a `sycl` component to `FindBLAS` rather than having a free-floating variable to manipulate these types of behaviors

ajaypanyala replied:
I tried that initially, could not get the `sycl` component to work. Will re-try and create an issue if it does not work this time.

wavefunction91 commented on FindIntelMKL.cmake:
```diff
@@ -296,7 +306,11 @@ if( "ilp64" IN_LIST IntelMKL_FIND_COMPONENTS )
+    set( IntelMKL_Fortran_COMPILE_FLAGS  "-fdefault-integer-8" )
   elseif( CMAKE_C_COMPILER_ID MATCHES "PGI" )
     set( IntelMKL_Fortran_COMPILE_FLAGS "-i8" )
   endif()
```
We may want to factor this out into a separate `Find` module (or something, not sure of the proper convention here) to handle integer based flags. Logic follows that these should also propagate to other BLAS distros in the case that ILP64 is requested

wavefunction91 commented on FindIntelMKL.cmake:
```diff
@@ -345,6 +359,10 @@ if( IntelMKL_LIBRARY AND IntelMKL_THREAD_LIBRARY AND IntelMKL_CORE_LIBRARY )
        ${IntelMKL_THREAD_LIBRARY} 
        ${IntelMKL_CORE_LIBRARY} )
 
+  if(ENABLE_DPCPP)
```
See previous comment on DPCPP

wavefunction91 commented on FindLAPACK.cmake:
```diff
@@ -44,7 +44,7 @@ if( NOT LAPACK_LIBRARIES )
   message( STATUS "LAPACK_LIBRARIES Not Given: Checking for LAPACK in BLAS" )
   set( LAPACK_LIBRARIES           ${BLAS_LIBRARIES}           )
-  set( LAPACK_INCLUDE_DIR         ${BLAS_INCLUDE_DIR}         )
+  set( LAPACK_INCLUDE_DIR         ${BLAS_INCLUDE_DIRS}        )
```
Change to `LAPACK_INCLUDE_DIRS` if that's going to be the convention

wavefunction91 commented on FindReferenceLAPACK.cmake:
```diff
@@ -46,6 +46,12 @@ else()
   set( ReferenceLAPACK_LIBRARIES ${ReferenceLAPACK_LP64_LIBRARIES} )
 endif()
 
+find_package(StandardFortran REQUIRED)
```
This can't be here, I have a module that checks for these things

https://github.com/wavefunction91/linalg-cmake-modules/blob/main/util/BLASUtilities.cmake#L43

Though I would like to see this `StandardFortran` module get used there

wavefunction91 commented on FindStandardFortran.cmake:
```diff
@@ -0,0 +1,41 @@
+elseif(CMAKE_Fortran_COMPILER_ID MATCHES "Flang")
+    set(STANDARDFORTRAN_LIBS flang flangrti pgmath)
+else()
+    message(FATAL_ERROR "${CMAKE_Fortran_COMPILER_ID} not supported yet.")
```
This need to be more descriptive in that its specifically in conjunction with this module

ajaypanyala committed:
find std. fortran libs

ajaypanyala committed:
sycl discovery as a blas component

ajaypanyala committed:
fix

ajaypanyala committed:
Find linalg dependencies during LAPACK, ScaLAPACK discovery

wavefunction91 committed:
Merge branch 'main' of https://github.com/ajaypanyala/linalg-cmake-modules into ajaypanyala-main

wavefunction91 committed:
Propagate sycl component in FindLAPACK

wavefunction91 commented on LinAlgModulesMacros.cmake:
```diff
@@ -1,14 +1,59 @@
 set( LINALG_MACROS_DIR ${CMAKE_CURRENT_LIST_DIR} )
 
+macro( find_linalg_dependencies _libs )
```
Would it still be correct if we checked just for `OpenMP` and `Threads`?

ajaypanyala replied:
Yes!

wavefunction91 commented on util/LAPACKUtilities.cmake:
```diff
@@ -37,16 +37,12 @@ foreach( _uplo LOWER UPPER )
+        include(CMakeFindDependencyMacro)
+        find_dependency(StandardFortran REQUIRED)
+        list( APPEND ${_libs} ${STANDARDFORTRAN_LIBRARIES} )
         set( ${_libs} ${${_libs}} PARENT_SCOPE )
```
This should only be appended if Fortran linkage is missing - it shouldn't be the default behaviour

wavefunction91 commented on util/ScaLAPACKUtilities.cmake:
```diff
@@ -37,16 +37,12 @@ foreach( _uplo LOWER UPPER )
+        find_dependency(StandardFortran REQUIRED)
+        list( APPEND ${_libs} ${STANDARDFORTRAN_LIBRARIES} )
         set( ${_libs} ${${_libs}} PARENT_SCOPE )
-
```
See above

wavefunction91 committed:
Starting CI

wavefunction91 committed:
Typo in CI script

wavefunction91 committed:
Sudo...

wavefunction91 committed:
Add OMP to build

wavefunction91 committed:
[CI] Explicitly disable NETLIB + Threads

wavefunction91 committed:
[CI] Added FindBLAS - Default Config to CI

wavefunction91 committed:
Typo

wavefunction91 committed:
[CI] Typo

wavefunction91 committed:
[CI] Add integer specific check in FindBLAS CI

wavefunction91 committed:
[CI] Add FindLAPACK

wavefunction91 committed:
[CI] Typo

wavefunction91 committed:
[CI] Minor

wavefunction91 committed:
[CI] Added integer interface checks for FindLAPACK CI

wavefunction91 committed:
[CI] Enable IntelMKL in CI matrix

wavefunction91 committed:
[CI] Fixup for MKL

wavefunction91 committed:
[CI] Slight refactor of CI matrix

wavefunction91 committed:
[CI] Add MPI + Move NETLIB over to ScaLAPACK

wavefunction91 committed:
[CI] Minor

wavefunction91 committed:
[CI] MPI fixup

wavefunction91 committed:
[CI] Sudo...

wavefunction91 committed:
[CI] Force failures on incorrect libraries for integer interface checks, print debug information in CI matrix

wavefunction91 committed:
[CI] env fixup

wavefunction91 committed:
[CI] Print debug information

wavefunction91 committed:
[CI] Fixup

wavefunction91 committed:
[CI] Minor

wavefunction91 committed:
[CI] Remove preinstalled BLAS/LAPACK

wavefunction91 committed:
[BLAS] Cache BLAS_LIBRARIES

wavefunction91 committed:
[CI] Typo

wavefunction91 committed:
[CI] Print out advanced variables

wavefunction91 committed:
[BLAS] Cache fixup

wavefunction91 committed:
[CI[ Add check that BLAS libraries are correct

wavefunction91 committed:
[CI] Added missing script

wavefunction91 committed:
[CI] Debug

wavefunction91 committed:
[CI] Change Sh -> Bash

wavefunction91 committed:
[CI] Try to force a failure

wavefunction91 committed:
[LAPACK/CI] Cache LAPACK_LIBRARIES + add test in CI

wavefunction91 committed:
[CI] Add ScaLAPACK checks

wavefunction91 committed:
[CI] Typo

wavefunction91 committed:
[CI] Update BLAS/LAPACK checks with integer types for MKL

wavefunction91 committed:
[CI] Disable ILP64 checks on default as LP64 is the default

wavefunction91 committed:
[CI] Debug...

wavefunction91 committed:
[CI] Debug...

wavefunction91 committed:
[CI] typo

wavefunction91 committed:
[CI] Fixup on disabled paths in CI matrix

wavefunction91 committed:
[CI] Improve MKL correctness checking + Add check for correct ScaLAPACK linkage

wavefunction91 committed:
[CI] Added missing script

wavefunction91 committed:
[CI] Added integer specific ScaLAPACK tests + fixup in CI matrix for Default ScaLAPACK + MKL-ILP64

wavefunction91 committed:
[CI] Add OpenMPI to CI matrix

wavefunction91 committed:
[CI] Minor

wavefunction91 committed:
[CI] Only test OpenMPI for MKL - trivial for others

wavefunction91 committed:
[CI] Try to convince CI matrix to run all MKL paths with OpenMPI

wavefunction91 committed:
[CI] Manual OpenMPI restriction

wavefunction91 committed:
Merge branch 'feature/ci' into ajaypanyala-main

wavefunction91 committed:
[MACROS] Minor

wavefunction91 committed:
[CI] Typo in CI Matrix

wavefunction91 committed:
Merge branch 'feature/ci' into ajaypanyala-main

wavefunction91 committed:
[BLAS/LAPACK] Cache more variables + add INCLUDE_DIRECTORIES/COMPILE_OPTIONS to LAPACK::LAPACK

wavefunction91 committed:
[ILP64] Added ILP64 check module

wavefunction91 committed:
[ILP64/IntelMKL] Move ILP64 flags from Intel MKL to FindBLAS/LAPACK/ScaLAPACK

wavefunction91 committed:
[IntelMKL] Cleanup of commented out ILP64 flags

wavefunction91 committed:
[MISC] Propagate XXX_PREFERS_STATIC and add script to check cmake cache in CI

wavefunction91 committed:
Merge remote-tracking branch 'ajay/main' into ajaypanyala-main

wavefunction91 committed:
[CI] Typo

wavefunction91 committed:
[CI] Change script to be executable

wavefunction91 committed:
[CI] Add output to cmake check script

wavefunction91 committed:
[CI] Added scripts to check LAPACK/ScaLAPACK CMake caches

wavefunction91 committed:
[CI] Add test for static linkage

wavefunction91 committed:
[CI] Added error output in CI matrix

wavefunction91 committed:
[BLIS/MACROS] Improve library appendage on missing symbols

wavefunction91 committed:
[MACROS] Factored out library appendange to separate function

wavefunction91 committed:
[MACROS] Factored library appendange into FindScaLAPACK

wavefunction91 committed:
[CI] Added static tests for FindLAPACK

wavefunction91 committed:
[LAPACK/ScaLAPACK] Copy metadeta in LAPACK/ScaLAPACK

wavefunction91 committed:
[MACROS] Cleanup

wavefunction91 committed:
[MISC] Added FindTBB from https://github.com/justusc/FindTBB, enabled TBB path in FindIntelMKL

wavefunction91 committed:
[CI] Enable TBB MKL paths in CI

wavefunction91 committed:
[CI] Typo

wavefunction91 committed:
[CI] Add TBB to build system in CI

wavefunction91 committed:
[CI] Typo

wavefunction91 committed:
[MACROS] Add TBB as a registered dependency

wavefunction91 committed:
[MISC] Cleanup debug print + add inclusion of dependency macro in LinAlgModules

wavefunction91 committed:
[BLAS] Added FindAccelerate and modified FindBLAS to look for it on Darwin systems

wavefunction91 committed:
[MKL] Added missing MKL_ILP64 compiler definition for MKL ILP64

wavefunction91 committed:
Merge branch 'main' of https://github.com/ajaypanyala/linalg-cmake-modules into main

ajaypanyala committed:
Cache LAPACK_IS_LP64

ajaypanyala committed:
Cache LP64 flags

ajaypanyala committed:
add MPI to find_linalg_dependencies

wavefunction91 committed:
[BLAS] Change BLAS LP64 check to analyze output rather than return code to solve problems with F* STOP return code

wavefunction91 committed:
[LAPACK] Change LAPACK LP64 check to analyze output rather than return code to solve problems with F* STOP return code

wavefunction91 committed:
[BLAS] Add CMakeFindDependencyMacro to FindBLAS for ILP64 check

wavefunction91 committed:
Add support of AOCL (AMD-BLIS/FLAME)

wavefunction91 committed:
[BLIS] Fix FindBLIS for AOCL

wavefunction91 committed:
[BLIS] Fix static BLIS discovery

ajaypanyala committed:
Additional check for missing Fortran symbols

ajaypanyala committed:
check compile output for sqrt when detecting missing libs

wavefunction91 committed:
Add functions to emulate Kitware variables BLA_STATIC and BLA_VENDOR

wavefunction91 committed:
Pin CI to Ubuntu-18.04

wavefunction91 committed:
Typo in CI

wavefunction91 committed:
More typos in CI

wavefunction91 committed:
Add BLA_VENDOR CI test

Commit:
Include missing standard macros in FindScaLAPACK

Commit:
Added IntelMKL TARGETs

evaleev committed:
Import Threads as global target (#1)

evaleev committed:
update findtbb (#2)

* typo

* replace FindTBB with the kitware version to be able to use with OneAPI MKL

this also supports using the targets installed via a cmake-configured install of TBB (and prefers such targets)

* explicitly link against C++ lib when using not only static but also pre-OneAPI _dynamic_ lib with TBB backend

* move dependence on c++ stdlib into TBB::tbb target

ajaypanyala committed:
fortran libs fix for ArmFlang

evaleev committed:
lapack linkage check uses `dgesvd` (#3)

* typo

* use dgesvd (instead of dpstrf) to detect linkage dependencies of lapack

on macos homebrew's openblas only some routines have dependence on fortran, dpstrf is not one of them

* typos

* consolidate check_{*}_exists into check_fortran_function_exists + for LAPACK test dpstrf AND dgesvd

* check_fortran_function_exists -> check_fortran_functions_exist to check for a set of functions at once

* detect silent compile errors in {blas,lapack}_check_int

* create genex-free linalg_<LANG>_COMPILE_OPTIONS that can be used in contexts (such as try_compile) where genex is not permitted

* factor unit tests

* do install util/BLASUtilities.cmake

wavefunction91 committed:
[CI] Force OpenBLAS GENERIC build in CI to avoid hangups

evaleev committed:
FORTRAN -> Fortran (#4)

* typo

* use dgesvd (instead of dpstrf) to detect linkage dependencies of lapack

on macos homebrew's openblas only some routines have dependence on fortran, dpstrf is not one of them

* typos

* consolidate check_{*}_exists into check_fortran_function_exists + for LAPACK test dpstrf AND dgesvd

* check_fortran_function_exists -> check_fortran_functions_exist to check for a set of functions at once

* detect silent compile errors in {blas,lapack}_check_int

* create genex-free linalg_<LANG>_COMPILE_OPTIONS that can be used in contexts (such as try_compile) where genex is not permitted

* factor unit tests

* do install util/BLASUtilities.cmake

* FORTRAN -> Fortran

* [CI] Force OpenBLAS GENERIC build in CI to avoid hangups

Co-authored-by: David Williams-Young <dbwy@lbl.gov>

wavefunction91 commented:
@ajaypanyala It's clear neither of us have the cycles to polish this to the extent that we would like. Maybe we should just merge and work out future needs/wants in other PRs?

ajaypanyala commented:
@wavefunction91 Sure. I do not have anything to add at this point too. Please go ahead and merge.
